core/src/storage2/lazy/lazy_hmap.rs

```diff
@@ -92,16 +93,39 @@ pub struct LazyHashMap<K, V, H> {
     hash_builder: RefCell<HashBuilder<H, Vec<u8>>>,
 }
 
+/// When querying `entry()` there is a case which needs special treatment:
+/// In `entry()` we first do a look-up in the cache. If the requested key is
+/// in the cache we return the found object.
+/// If it is not in the cache we query the storage. If we find the element
+/// in storage we insert it into the cache.
+///
+/// The problem now is that in this case we only have the `Vacant` object
+/// which we got from searching in the cache, but we need to return an
+/// `Occupied` here, since the object is now in the cache. We could do this
+/// by querying the cache another time -- but this would be an additional
+/// search. So what we do instead is to save a reference to the inserted
+/// cache value in the `Occupied`. As a consequence all Entry API operations
+/// (`get`, `remove`, ...) need to distinguish both cases.
+enum EntryOrMutableValue<E, V> {
+    /// An occupied `EntryMap` entry that holds a value.
+    Entry(E),
+    /// A reference to the mutable value behind a cache entry.
+    MutableValue(V),
```

Comment:
The name should make it more clear that this represents the case where the element was not in the cache but in the storage.